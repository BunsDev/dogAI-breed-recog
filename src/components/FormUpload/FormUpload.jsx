import React, { useState, useCallback } from 'react'
import { useDropzone } from 'react-dropzone'
import * as tmImage from '@teachablemachine/image'

import './FormUpload.css'

import uploadImage from '../../assets/uploadImage.svg'
import close from '../../assets/close.svg'

import PredictionResult from '../PredictionResult/PredictionResult.jsx'
import Loading from '../Loading/Loading.jsx'
import Preview from '../Preview/Preview.jsx'

function FormUpload() {
  const [predictions, setPredictions] = useState([])
  const [file, setFiles] = useState([])
  const [loading, setLoading] = useState(true)
  const [error, setError] = useState(null)
  const [description, setDescription] = useState({})

  const onDrop = useCallback(async acceptedFiles => {
    try {
      const start = Date.now()
      setError(null)
      setLoading(true)
      setPredictions([])
      setDescription({})

      const singleFile = acceptedFiles[0]
      setFiles(
        Object.assign(singleFile, {
          preview: URL.createObjectURL(singleFile)
        })
      )

      const image = document.getElementById('image')

      const model = await tmImage.load(
        './model/model.json',
        './model/metadata.json'
      )

      const prediction = await model.predictTopK(image, 3)
      handleResults(prediction)

      async function handleResults(data) {
        setPredictions(data)
        let breed = data[0].className.replace(/(_)/gi, ' ')
        if (breed === 'Cão Selvagem') {
          breed = 'Canídeos'
        }
        const wikipediaApiUrl = encodeURI(
          `https://pt.wikipedia.org/w/api.php?origin=*&action=opensearch&search=${breed}&limit=1`
        )
        try {
          const response = await fetch(wikipediaApiUrl)
          const responseData = await response.json()
          const wiki = {
            desc: responseData[2][0],
            wikiUrl: responseData[3][0]
          }
          if (!responseData[1].length) {
            wiki.error = true
          }
          setDescription(wiki)
        } catch (error) {
          console.error(error)
        }
      }

      setLoading(false)
      const end = new Date()
      const elapsedTime = end - start
      console.log(`It took ${elapsedTime / 1000} seconds`)
    } catch (err) {
      setLoading(false)
      setError(err)
      console.error(err)
    }
  }, [])

  const {
    getRootProps,
    getInputProps,
    isDragActive,
    isDragReject
  } = useDropzone({ onDrop, accept: 'image/jpeg, image/png, image/jpg' })

  const renderDragMessage = (isDragActive, isDragReject) => {
    if (!isDragActive) {
      return (
        <div className="dropContainer">
          <img src={uploadImage} alt="Illustration upload" />
          <h3>Arraste e solte ou clique aqui</h3>
          <small>para enviar sua imagem</small>
          <p>Nenhum arquivo selecionado</p>
        </div>
      )
    }
    if (isDragReject) {
      return (
        <div className="dropContainer dragReject">
          <img src={close} alt="X icon" />
          <h3>Arquivo não suportado!</h3>
        </div>
      )
    }

    return (
      <div className="dropContainer dragActive">
        <img src={uploadImage} alt="Illustration upload" />
        <h3>Solte aqui</h3>
        <small>para enviar sua imagem</small>
      </div>
    )
  }

  return (
    <>
      {file.name ? (
        <>
          <section>
            <section className="fileInput">
              <button {...getRootProps()} className="btn">
                Enviar outra imagem
              </button>
              <input {...getInputProps()} />
              <Preview f={file} />

              {loading && <Loading />}

              {error && (
                <p className="alert is-danger">
                  Erro ao efetuar análise, consulte os logs
                </p>
              )}
            </section>

            {predictions.length !== 0 && (
              <PredictionResult
                predictions={predictions}
                description={description}
              />
            )}
          </section>
        </>
      ) : (
        <div {...getRootProps()}>
          <input {...getInputProps()} />
          <h2>Selecione uma imagem</h2>
          {renderDragMessage(isDragActive, isDragReject)}
        </div>
      )}
    </>
  )
}

export default FormUpload
