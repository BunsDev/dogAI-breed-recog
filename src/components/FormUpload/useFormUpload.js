import { useCallback, useState } from 'react'
import * as tmImage from '@teachablemachine/image'

export function useFormUpload() {
  const [file, setFile] = useState([])
  const [prediction, setPrediction] = useState([])
  const [description, setDescription] = useState({})
  const [error, setError] = useState(null)

  const statusType = {
    IDLE: 'IDLE',
    FULFILLED: 'FULFILLED',
    REJECTED: 'REJECTED',
  }

  const [status, setStatus] = useState(statusType.IDLE)

  const handleResults = async (data) => {
    setPrediction(data)

    let breed = data.className.replace(/(_)/gi, ' ')
    if (breed === 'CÃ£o Selvagem') {
      breed = 'Canids'
    }

    const wikipediaApiUrl = encodeURI(
      `${process.env.REACT_APP_WIKIPEDIA_ENDPOINT}${breed}`
    )

    try {
      const response = await fetch(wikipediaApiUrl)

      if (response.status === 200 && response.ok) {
        let filteredDesc
        const wikipediaPages = await response.json()

        if (wikipediaPages.query) {
          wikipediaPages.query.pages.forEach((page) => {
            if (
              page.title.toLowerCase().includes(breed) &&
              page.extract.search(/(dog|canid|breed)/g) !== -1
            ) {
              filteredDesc = page.extract
            } else {
              filteredDesc = wikipediaPages.query.pages[0].extract
            }
          })
          setDescription({
            desc: `${filteredDesc.substring(0, 400 - 10)}...`,
            wikiUrl: `${process.env.REACT_APP_WIKIPEDIA_WIKI}/${breed}`,
          })
        } else {
          setDescription({
            desc: 'No wikipedia found.',
            wikiUrl: '',
            error: true,
          })
        }
      }
    } catch (err) {
      throw new Error(err)
    }
  }

  const onDrop = useCallback(async (acceptedFiles) => {
    try {
      const [single] = acceptedFiles
      setFile(
        Object.assign(single, {
          preview: URL.createObjectURL(single),
        })
      )

      // Load model and pick the best prediction out of 10 classes

      const image = document.getElementById('image')
      const model = await tmImage.load(
        './model/model.json',
        './model/metadata.json'
      )
      const predictions = await model.predictTopK(image, 10)
      const highestPrediction = predictions.reduce(
        (prev, current) =>
          prev.probability > current.probability ? prev : current,
        0
      )

      await handleResults(highestPrediction)
    } catch (err) {
      setError(err)
      setStatus(statusType.REJECTED)
    } finally {
      setStatus(statusType.FULFILLED)
    }
  }, [])

  return {
    onDrop,
    file,
    prediction,
    description,
    error,
    isLoading: status === statusType.IDLE,
    isFinished: status === statusType.FULFILLED,
    isRejected: status === statusType.REJECTED,
  }
}
