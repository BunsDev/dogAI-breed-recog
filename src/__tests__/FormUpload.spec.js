import React from 'react'
import { create } from 'react-test-renderer'
import FormUpload from '../components/FormUpload/FormUpload.jsx'

describe('FormUpload component', () => {
  test('Matches the snapshot', () => {
    const component = create(<FormUpload />)
    expect(component).toMatchSnapshot()
  })
})
