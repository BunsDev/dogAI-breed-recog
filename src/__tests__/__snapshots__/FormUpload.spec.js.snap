// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormUpload component Matches the snapshot 1`] = `
<div
  onBlur={[Function]}
  onClick={[Function]}
  onDragEnter={[Function]}
  onDragLeave={[Function]}
  onDragOver={[Function]}
  onDrop={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  tabIndex={0}
>
  <input
    accept="image/jpeg, image/png, image/jpg"
    autoComplete="off"
    multiple={true}
    onChange={[Function]}
    onClick={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    tabIndex={-1}
    type="file"
  />
  <h2>
    Selecione uma imagem
  </h2>
  <div
    className="dropContainer"
  >
    <img
      alt="Illustration upload"
      src="uploadImage.svg"
    />
    <h3>
      Arraste e solte ou clique aqui
    </h3>
    <small>
      para enviar sua imagem
    </small>
    <p>
      Nenhum arquivo selecionado
    </p>
  </div>
</div>
`;
